src/py/pyodide/console.py

```diff
@@ -39,43 +40,102 @@ class redirect_stdin(_RedirectStream[Any]):
-        self, write_handler: Callable[[str], Any], name: str | None = None
-    ) -> None:
-        self.write_handler = write_handler
-        self.name = name
+    @property
+    def encoding(self):
+        return self._encoding
 
-    def write(self, text: str) -> None:
-        self.write_handler(text)
+    @property
+    def errors(self):
+        return self._errors
 
-    def flush(self) -> None:
-        pass
+    @property
+    def name(self):
+        return self._name
 
-    def isatty(self) -> bool:
+    def isatty(self):
         return True
 
 
-class _ReadStream:
-    """A utility class so we can specify our own handler for reading from stdin"""
-
+class _WriteStream(_Stream):
     def __init__(
-        self, read_handler: Callable[[int], str], name: str | None = None
-    ) -> None:
-        self.read_handler = read_handler
-        self.name = name
+        self,
+        write_handler: Callable[[str], int | None],
+        name: str,
+        encoding: str = "utf-8",
+        errors: str = "strict",
+    ):
+        super().__init__(name, encoding, errors)
+        self._write_handler = write_handler
 
-    def readline(self, n: int = -1) -> str:
-        return self.read_handler(n)
+    def writable(self) -> bool:
+        return True
 
-    def flush(self) -> None:
-        pass
+    def write(self, s: str) -> int:
+        if self.closed:
+            raise ValueError("write to closed file")
+        s = str.encode(s, self.encoding, self.errors).decode(self.encoding, self.errors)
+        written = self._write_handler(s)
+        if written is None:
+            # They didn't tell us how much they wrote, assume it was the whole string
+            return len(s)
+        return written
 
-    def isatty(self) -> bool:
+
+class _ReadStream(_Stream):
+    def __init__(
+        self,
+        read_handler: Callable[[int], str],
+        name: str,
+        encoding: str = "utf-8",
+        errors: str = "strict",
+    ):
+        super().__init__(name, encoding, errors)
+        self._read_handler = read_handler
+        self._buffer = ""
+
+    def readable(self) -> bool:
         return True
 
+    def read(self, size: int | None = -1) -> str:
+        if self.closed:
+            raise ValueError("read from closed file")
+        if size is None:
+            # For some reason sys.stdin.read(None) works, but
+            # sys.stdin.readline(None) raises a TypeError
+            size = -1
+        if not isinstance(size, int):
+            raise TypeError(
+                f"argument should be integer or None, not '{type(size).__name__}'"
+            )
+        if 0 <= size < len(self._buffer):
+            result = self._buffer[:size]
+            self._buffer = self._buffer[size:]
+            return result
+        if size >= 0:
+            size -= len(self._buffer)
+        result = self._buffer
+        got = self._read_handler(size)
+        self._buffer = got[size:]
+        return result + got[:size]
```

Comment:
If the intention is to read all buffer, shouldn't this be

```py
if size >= 0:
    size = len(self._buffer)
```

?

Reply:
Well we need to call `_read_handler` with how much more we need. If we've been asked for a positive amount which is within the range of what we already have in the buffer, we just return a chunk of the buffer. If not, we've either been given a negative size (meaning read as much as possible) or a size larger than the buffer (meaning we have to make an onward call to read only the remainder).